benches/lua.rs

```diff
@@ -9,167 +8,121 @@ use vector::{
-                },
-            )
-        })
-        .with_function("v1", move |b| {
-            b.iter_with_setup(
-                || {
-                    let source = format!("event['{}'] = '{}'", key, value);
-                    (
-                        Event::new_empty_log(),
-                        transforms::lua::v1::Lua::new(&source, vec![]).unwrap(),
-                    )
-                },
-                |(event, mut transform)| {
-                    let event = transform.transform(event).unwrap();
-                    assert_eq!(event.as_log()[key], value_bytes_v1);
-                },
-            )
-        })
-        .with_function("v2", move |b| {
-            b.iter_with_setup(
-                || {
-                    let config = format!(
-                        r#"
-                        hooks.process = """
-                            function (event, emit)
-                                event.log['{}'] = '{}'
+    let mut benchmarks = [
+        ("native", {
+            let mut map = IndexMap::new();
+            map.insert(String::from(key), value.to_owned().into());
+            Box::new(transforms::add_fields::AddFields::new(map, true).unwrap())
+                as Box<dyn Transform>
+        }),
+        ("v1", {
+            let source = format!("event['{}'] = '{}'", key, value);
 
-                                emit(event)
-                            end
-                        """
-                        "#,
-                        key, value
-                    );
-                    (
-                        Event::new_empty_log(),
-                        transforms::lua::v2::Lua::new(
-                            &toml::from_str::<LuaConfig>(&config).unwrap(),
-                        )
-                        .unwrap(),
-                    )
-                },
-                |(event, mut transform)| {
-                    let event = transform.transform(event).unwrap();
-                    assert_eq!(event.as_log()[key], value_bytes_v2);
-                },
-            )
-        })
-        .throughput(Throughput::Elements(1)),
-    );
+            Box::new(transforms::lua::v1::Lua::new(&source, vec![]).unwrap()) as Box<dyn Transform>
+        }),
+        ("v2", {
+            let config = format!(
+                r#"
+    hooks.process = """
+        function (event, emit)
+            event.log['{}'] = '{}'
+
+            emit(event)
+        end
+    """
+    "#,
+                key, value
+            );
+            Box::new(
+                transforms::lua::v2::Lua::new(&toml::from_str::<LuaConfig>(&config).unwrap())
+                    .unwrap(),
+            ) as Box<dyn Transform>
+        }),
+    ];
+
+    for (name, transform) in benchmarks.iter_mut() {
+        group.bench_function(name.to_owned(), |b| {
+            b.iter(|| {
+                let event = black_box(transform.transform(event.clone()).unwrap());
```

Comment:
I added these `black_box` to ensure that the compiler doesn't optimize away the benchmark due to it discovering that the value is never actually used (except in debug mode where there is an assertion).

criterion automatically wraps inputs when `bench_with_input` is used and appears to wrap outputs of the iter closure. Instead of `black_box` I could write this as:

```
                 let event = transform.transform(event.clone()).unwrap();
                 debug_assert_eq!(event.as_log()[key], value.to_owned().into());
                 event
```

To return the event and have criterion wrap it in `black_box`.